### PR TITLE
fix(kubernetes): add support for json patch #3418

### PR DESCRIPTION
**Pull request by maggieneterval**
Closes #3418
- Adds JSON editor component
- Patch body editor toggles between YAML and JSON depending on selected merge strategy, and converts existing input between YAML/JSON appropriately
- See corresponding [CloudDriver change](https://github.com/spinnaker/clouddriver/pull/3047).

json patch

### CODE REVIEW

File: app/scripts/modules/kubernetes/src/v2/manifest/json/JsonEditor.tsx [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+import AceEditor from 'react-ace';
+import { $log } from 'ngimport';
+
+import 'brace/theme/textmate';
+import 'brace/mode/json';
+import './jsonEditor.less';
+
+export interface IJsonEditorProps {
+  value: string;
+  onChange(raw: string, manifest: any): void;
```
Comment (danielpeach) [inline app/scripts/modules/kubernetes/src/v2/manifest/json/JsonEditor.tsx]: Using the word `manifest` here for the yaml editor was a mistake on my part - maybe `obj` or something? (Doesn't seem like this component necessarily has to be used for editing manifests)

File: app/scripts/modules/kubernetes/src/v2/manifest/yaml/yamlEditor.less [hunk 2]
```diff
@@ -1,6 +1,6 @@
 #yaml-editor {
```
Comment (danielpeach) [inline app/scripts/modules/kubernetes/src/v2/manifest/yaml/yamlEditor.less]: maybe this file could be renamed and the selectors could apply to both editors?

File: app/scripts/modules/kubernetes/src/v2/manifest/json/jsonEditor.component.ts [hunk 3]
```diff
@@ -0,0 +1,7 @@
+import { module } from 'angular';
+import { react2angular } from 'react2angular';
+
+import { JsonEditor } from './JsonEditor';
+
+export const JSON_EDITOR_COMPONENT = 'spinnaker.core.jsonEditor.component';
```
Comment (danielpeach) [inline app/scripts/modules/kubernetes/src/v2/manifest/json/jsonEditor.component.ts]: s/`spinnaker.core.jsonEditor.component`/`spinnaker.kubernetes.jsonEditor.component`

File: app/scripts/modules/kubernetes/src/v2/pipelines/stages/patchManifest/patchManifestConfig.controller.ts [hunk 4]
```diff
@@ -84,4 +97,42 @@ export class KubernetesV2PatchManifestConfigCtrl implements IController {
   public handleManifestSelectorChange = (): void => {
     this.$scope.$applyAsync();
   };
+
+  public getMergeStrategy = (): MergeStrategy => {
```
Comment (danielpeach) [inline app/scripts/modules/kubernetes/src/v2/pipelines/stages/patchManifest/patchManifestConfig.controller.ts]: if this method is only called within this controller, the access modifier can be `private`

File: app/scripts/modules/kubernetes/src/v2/pipelines/stages/patchManifest/patchManifestConfig.controller.ts [hunk 5]
```diff
@@ -84,4 +97,42 @@ export class KubernetesV2PatchManifestConfigCtrl implements IController {
...
+    return this.$scope.stage.options.mergeStrategy;
+  };
+
+  public getEditorMode = (): EditorMode => {
+    return get(mergeStrategyToEditorMode, this.getMergeStrategy());
+  };
+
+  public setRawPatchBody = (mergeStrategy: MergeStrategy, patchBody: any): void => {
```
Comment (danielpeach) [inline app/scripts/modules/kubernetes/src/v2/pipelines/stages/patchManifest/patchManifestConfig.controller.ts]: same here re: access modifier

File: app/scripts/modules/kubernetes/src/v2/pipelines/stages/patchManifest/patchManifestConfig.controller.ts [hunk 6]
```diff
@@ -84,4 +97,42 @@ export class KubernetesV2PatchManifestConfigCtrl implements IController {
   public handleManifestSelectorChange = (): void => {
     this.$scope.$applyAsync();
   };
+
+  public getMergeStrategy = (): MergeStrategy => {
+    return this.$scope.stage.options.mergeStrategy;
+  };
+
+  public getEditorMode = (): EditorMode => {
+    return get(mergeStrategyToEditorMode, this.getMergeStrategy());
```
Comment (danielpeach) [inline app/scripts/modules/kubernetes/src/v2/pipelines/stages/patchManifest/patchManifestConfig.controller.ts]: I think you could get away with not using `get` here

File: app/scripts/modules/kubernetes/src/v2/manifest/json/JsonEditor.tsx (hunk 1 above)
Comment (maggieneterval) [reply app/scripts/modules/kubernetes/src/v2/manifest/json/JsonEditor.tsx]: Good call, replaced with `obj` here and in the `YamlEditor`.

File: app/scripts/modules/kubernetes/src/v2/manifest/yaml/yamlEditor.less (hunk 2 above)
Comment (maggieneterval) [reply app/scripts/modules/kubernetes/src/v2/manifest/yaml/yamlEditor.less]: Done!

**Pull request merged**